waku/v2/peermanager/peer_manager.go

```diff
@@ -120,3 +109,86 @@ func (pm *PeerManager) pruneInRelayConns() {
+	info, err := peer.AddrInfoFromP2pAddr(address)
+	if err != nil {
+		return "", err
+	}
+
+	//Add Service peers to serviceSlots.
+	for _, service := range protocols {
+		if service != WakuRelayIDv200 {
+			pm.AddPeerToServiceSlot(service, info.ID)
+		}
+	}
+	//Set origin, ENR and Direction and any other required items in Waku Peer Store.
+	pm.logger.Info("adding peer to peerstore", logging.HostID("peer", info.ID))
+	err = pm.host.Peerstore().(wps.WakuPeerstore).SetOrigin(info.ID, origin)
+	if err != nil {
+		return "", err
+	}
+
+	pm.host.Peerstore().AddAddrs(info.ID, info.Addrs, peerstore.AddressTTL)
+	err = pm.host.Peerstore().AddProtocols(info.ID, protocols...)
+	if err != nil {
+		return "", err
+	}
+	return info.ID, nil
+}
+
+// RemovePeer deletes peer from the peerStore after disconnecting it.
+func (pm *PeerManager) RemovePeer(peerID peer.ID) {
+	// TODO: Need to handle removePeer also via peermanager
+	// Need to updated serviceSlot accordingly.
+
+}
+
+// AddServicePeer adds a peerID to serviceSlot.
+// Adding to peerStore is expected to be already done by caller.
+// If relay proto is passed, it is not added to serviceSlot.
+func (pm *PeerManager) AddPeerToServiceSlot(proto protocol.ID, peerID peer.ID) {
+	if proto == WakuRelayIDv200 {
+		pm.logger.Warn("Cannot add Relay peer to service peer slots")
+		return
+	}
+	//For now adding the peer to serviceSlot which means the latest added peer would be given priority.
+	//TODO: Ideally we should maintain multiple peers per service and return best peer based on peer score or RTT etc.
+	//Should we override service slot if peer is identified dynamically?
+	pm.logger.Info("Adding peer to service slots", zap.String("peerId", peerID.Pretty()), zap.String("service", string(proto)))
+	pm.serviceSlots[proto] = peerID
+}
+
+// SelectPeer is used to return a random peer that supports a given protocol.
+// If a list of specific peers is passed, the peer will be chosen from that list assuming
+// it supports the chosen protocol, otherwise it will chose a peer from the service slot.
+// If a peer cannot be found in the service slot, a peer will be selected from node peerstore
+func (pm *PeerManager) SelectPeer(proto protocol.ID, specificPeers []peer.ID, logger *zap.Logger) (peer.ID, error) {
+	// @TODO We need to be more strategic about which peers we dial. Right now we just set one on the service.
+	// Ideally depending on the query and our set  of peers we take a subset of ideal peers.
+	// This will require us to check for various factors such as:
+	//  - which topics they track
+	//  - latency?
+
+	filteredPeers, err := utils.FilterPeersByProto(pm.host, specificPeers, proto)
+	if err != nil {
+		return "", err
+	}
+	if proto == WakuRelayIDv200 {
+		if filteredPeers.Len() > 0 {
+			pm.logger.Info("Got peer from peerstore", zap.String("peerId", filteredPeers[0].Pretty()))
+			return filteredPeers[0], nil
```

Comment:
Should a random peer be returned instead to match the function description?

Reply:
Good catch!